handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -1099,76 +1117,26 @@ private SSLEngineResult unwrap(
-                                case 0:
-                                    out.ensureWritable(Math.min(max, in.readableBytes()));
-                                    break;
-                                default:
-                                    out.ensureWritable(max);
-                            }
-                            break;
-                        default:
-                            return result;
-                    }
-                }
+                singleBuffer[0] = out.nioBufferCount() == 1 ? out.internalNioBuffer(writerIndex, out.writableBytes()) :
+                        out.nioBuffer(writerIndex, out.writableBytes());
+                SSLEngineResult result = opensslEngine.unwrap(in0, singleBuffer);
+                out.writerIndex(writerIndex + result.bytesProduced());
+                return result;
             } finally {
                 singleBuffer[0] = null;
             }
         } else {
-            int overflows = 0;
-            ByteBuffer in0;
-            if (nioBufferCount == 1) {
-                // Use internalNioBuffer to reduce object creation.
-                in0 = in.internalNioBuffer(readerIndex, len);
-            } else {
-                // This should never be true as this is only the case when OpenSslEngine is used, anyway lets
-                // guard against it.
-                in0 = in.nioBuffer(readerIndex, len);
-            }
-            for (;;) {
-                int writerIndex = out.writerIndex();
-                int writableBytes = out.writableBytes();
-                ByteBuffer out0;
-                if (out.nioBufferCount() == 1) {
-                    out0 = out.internalNioBuffer(writerIndex, writableBytes);
-                } else {
-                    out0 = out.nioBuffer(writerIndex, writableBytes);
-                }
-                SSLEngineResult result = engine.unwrap(in0, out0);
-                out.writerIndex(out.writerIndex() + result.bytesProduced());
-                switch (result.getStatus()) {
-                    case BUFFER_OVERFLOW:
-                        int max = engine.getSession().getApplicationBufferSize();
-                        switch (overflows ++) {
-                            case 0:
-                                out.ensureWritable(Math.min(max, in.readableBytes()));
-                                break;
-                            default:
-                                out.ensureWritable(max);
-                        }
-                        break;
-                    default:
-                        return result;
-                }
-            }
+            ByteBuffer in0 = nioBufferCount == 1 ? in.internalNioBuffer(readerIndex, len) :
+                    in.nioBuffer(readerIndex, len);
+
+            ByteBuffer out0 = out.nioBufferCount() == 1 ? out.internalNioBuffer(writerIndex, out.writableBytes()) :
+                    out.nioBuffer(writerIndex, out.writableBytes());
```

Comment:
You could extract this into a static method, because you do this 3 times in this change.